### ISSUE TITLE
Performance issue with complex function

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3082]
This is a tracking bug for an issue that was submitted by @davrot in the pylance-release project.

Original issue thread: https://github.com/microsoft/pylance-release/issues/2299

**Comment by davrot** [comment]
Yesterday, I was trying to isolate the corresponding line but wasn't able to.
I added an exit(1) at the beginning the source code and then I was moving it slowly down to the end of the source code.
As long as the exit(1) was in there, the syntax highlighting works (after restarting of the language server).
To be precise: The part of the source code before the exit(1) is processed and the part after the exit(1) is faded out as expected. And no 100% cpu.
The language server doesn't have problems until the last lines. see the pictures okay and broken. As reference see the picture... well.. reference for what the indents are for.

Then I thought I could replace the content after
with torch.no_grad():
    while True:
        for h_x, h_x_labels in my_loader_train:

with pass but then the error is different. It does just ignores the last line (tb.close()) and doesn't stay at 100%. Strange! Here I gave up...  



![reference](https://user-images.githubusercontent.com/54365609/154808687-5393e4a3-ca4d-4b3f-b7ff-e13f4c3b43a7.png)
![broken](https://user-images.githubusercontent.com/54365609/154808689-d9943301-15f3-4dde-93ff-d20295c05b66.png)
![okay](https://user-images.githubusercontent.com/54365609/154808690-934a2fd8-b057-4ab0-af82-29aef63aa096.png)

**Comment by debonte** [comment]
I found that commenting out either of these two blocks seems to "fix" the issue:

```python
reconstuction = (
    np.expand_dims(output_h, axis=(1, 2, 3))
    * np.expand_dims(wf, axis=(0, 5, 6))
).sum(axis=4)
```

```python
wf, bad_temp = update_weight_forward(
    wf,
    d_w_temp,
    loss_f_lp_max,
    loss_f_lp,
    learning_rate_gamma_w,
    learning_rate_threshold_w,
)
```

**Comment by erictraut** [comment]
Yes, the problem is related to a complex cyclical dependency between `wf` and other variables in the loop. There are over 80 local variables (!!!) in this scope. None of them contain type annotations, so their types all need to be inferred based on assignments. Many of them are modified within the inner loop, and their types depend on the types of others in a circular manner. This is resulting in a combinatoric explosion. To make matters worse, many of the function calls to numpy functions are overloads, and those require expensive speculative evaluation to determine the relevant overload. 

I don't know if I'm going to be able to find a way for the type analyzer to perform well on this code. As a fallback, I may need to look for ways to detect this type of complexity and either "give up" or don't attempt analysis if the complexity level is beyond some threshold. This is the approach the TypeScript compiler takes when it encounters code with high complexity. The problem is compounded in unannotated Python code because the type of every symbol needs to be inferred.

**Comment by davrot** [comment]
I tested it with annotated variables and then it works (see the appended version). I takes ~ 2 sec or so. Personally, I would prefer the error "your code is too complex; please use annotations.". :-) Then I can ignore it and can use the language server for the other files without restarting it.
 
However, since I don't get any other benefits from annotating the code in the moment, I will ignore the error. :-) mypy gives me false positives in the torch.tensor( & jumping between torch.tensor and np.numpy) part... 


[learn_it_mnist_local_many_ano.py.txt](https://github.com/microsoft/pyright/files/8102941/learn_it_mnist_local_many_ano.py.txt)

**Comment by davrot** [comment]
I was able to trace the problem to ONE line. If I annotate it and leave the rest unannotated then it works. 

Broken:
            reconstuction = (
                np.expand_dims(output_h, axis=(1, 2, 3))
                * np.expand_dims(wf, axis=(0, 5, 6))
            ).sum(axis=4)

Okay:
            reconstuction: np.ndarray = (
                np.expand_dims(output_h, axis=(1, 2, 3))
                * np.expand_dims(wf, axis=(0, 5, 6))
            ).sum(axis=4)

**Comment by davrot** [comment]
I went through the whole code base of this project and annotated every file of those down to every variable, class, method, and function. mypy is now totally happy:
[admin@fedora LocalLearning]$ mypy *.py 
Success: no issues found in 33 source files

However, I found this py file which drives Code into a cpu frenzy from which it never (means: I waited several minutes :-) ) comes back. 

Well, it is even more complex but this is really on the low complex side for our research... 

[learn_dual_last_layer_mnist_local_many.txt](https://github.com/microsoft/pyright/files/8116074/learn_dual_last_layer_mnist_local_many.txt)
![2022-02-22_12-13](https://user-images.githubusercontent.com/54365609/155121331-f4bdfdbe-9c03-497f-8ba2-a3e0413d5172.png)

**Comment by erictraut** [comment]
I found and fixed the root cause of the slow analysis time for this code. The fix will be included in the next release of pyright and pylance.

**Comment by erictraut** [comment]
This is addressed in 1.1.226, which I just published. It will also be included in the next release of pylance.
